pyro/infer/predictive.py

```diff
@@ -255,36 +268,25 @@ def log_like_helper(*args, **kwargs) -> torch.Tensor:
             for name, site in trace.nodes.items()
             if site["type"] == "sample" and site["is_observed"]
         }
-        # now trace it and extract the likelihood from observed sites
-        predictive = Predictive(
-            model, posterior_samples, guide, num_samples, (), parallel)
         # use vectorized trace
         if parallel:
             log_like = dict()
+            # now trace it and extract the likelihood from observed sites
+            predictive = Predictive(
+                model, posterior_samples, guide, num_samples, (), parallel)
             trace = predictive.get_vectorized_trace(*args, **kwargs)
             for obs_name, obs_val in observed.items():
                 obs_site = trace.nodes[obs_name]
                 log_like[obs_name] = obs_site["fn"].log_prob(obs_val).detach().cpu()
         # iterate over samples from posterior if model can't be vectorized
         else:
-            all_samples = {
-                k: v for k, v in predictive(*args, **kwargs).items()
-                if k not in observed}
-            log_like = {obs_name: [] for obs_name in observed}
-            for i in range(num_samples):
-                param_sample = {k: v[i] for k, v in all_samples.items()}
-                cond_model = pyro.condition(model, data=param_sample)
-                traced_model = poutine.trace(cond_model)
-                trace = traced_model.get_trace(*args, **kwargs)
-                for obs_name, obs_val in observed.items():
-                    obs_site = trace.nodes[obs_name]
-                    ll_sample = obs_site["fn"].log_prob(obs_val).detach().cpu()
-                    log_like[obs_name].append(ll_sample)
-            # each sample is a tensor in a list; stack all the elements into
-            # a single tensor for each site
-            log_like = {
-                obs_name: torch.stack(ll_list, dim=0)
-                for obs_name, ll_list in log_like.items()
+            site_shapes = {
+                name: (num_samples,) + trace.nodes[name]["value"].shape
```

Comment:
I think we need to use `name: (num_samples,) + trace.nodes[name]["fn"].batch_shape` or to make things safe, we need to broadcast `fn.batch_shape` and `value.shape[:len(value.shape) - fn.event_dim]`, where `value = trace.nodes[name]["value"]` and `fn = trace.nodes[name]["fn"]`. We need to add some smoke test for likelihoods with `event_dim != 0` (e.g. MultivariateNormal or dist.Normal(...).to_event(1)) to catch this issue.

Reply:
Does [this test](https://github.com/pyro-ppl/pyro/pull/2684/files#diff-8d729ab9d44de7289fe0dd334f312a5257b0f03ca91ccf1a5b6b549fdb6a3c1eR161-R181) cover that possibility? It seems like I didn't need to make use of the broadcasting trick to get it to work.

Reply:
Yes, it does. Thanks! Broadcasting is only useful in case
```
pyro.sample("a", dist.Normal(0, 1).expand([10]), obs=torch.tensor(1.))
```
like in the following program
```python
def model():
    x = pyro.sample("x", dist.Normal(0, 1).expand([10]))
    pyro.sample("obs", dist.Normal(x, 1), obs=torch.tensor(0.))

mcmc = MCMC(NUTS(model), 1000)
mcmc.run()
```

Reply:
oh, I see what you mean now

Reply:
I added [some test](https://github.com/pyro-ppl/pyro/pull/2684/commits/52adf13cb9bfd61290fb9bb5d6efbdc9ae83fc8d#diff-8d729ab9d44de7289fe0dd334f312a5257b0f03ca91ccf1a5b6b549fdb6a3c1eR196-R216) to verify broadcasting works; it seems to work naturally for the `Normal` distribution.